main.py

```diff
@@ -1272,6 +1291,55 @@ def tearDown(self):
             p.strip_dirs().sort_stats("tottime").print_stats(0.05)
 
 
+class Test_Index(Timing, unittest.TestCase):
+    def check_idx2_reverse(self, ij):
+        i, j = compound_idx2_reverse(ij)
+        self.assertTrue(i <= j)
+        self.assertEqual(ij, compound_idx2(i, j))
+
+    def check_idx4_reverse(self, ijkl):
+        i, j, k, l = compound_idx4_reverse(ijkl)
+        ik = compound_idx2(i, k)
+        jl = compound_idx2(j, l)
+        self.assertTrue(i <= k)
+        self.assertTrue(j <= l)
+        self.assertTrue(ik <= jl)
+        self.assertEqual(ijkl, compound_idx4(i, j, k, l))
+
+    def check_idx4_reverse_all(self, ijkl):
+        for i, j, k, l in compound_idx4_reverse_all(ijkl):
+            self.assertEqual(compound_idx4(i, j, k, l), ijkl)
+
+    def check_canonical_idx4(self, ijkl):
+        for i, j, k, l in compound_idx4_reverse_all(ijkl):
+            self.assertEqual(
+                canonical_idx4(*compound_idx4_reverse(ijkl)), canonical_idx4(i, j, k, l)
+            )
+
+    def check_compound_idx4_reverse_is_canonical(self, ijkl):
+        self.assertEqual(compound_idx4_reverse(ijkl), canonical_idx4(*compound_idx4_reverse(ijkl)))
+
+    def test_idx2_reverse(self, n=10000, nmax=(1 << 63) - 1):
+        for ij in random.sample(range(nmax), k=n):
+            self.check_idx2_reverse(ij)
+
+    def test_idx4_reverse(self, n=10000, nmax=(1 << 63) - 1):
+        for ijkl in random.sample(range(nmax), k=n):
+            self.check_idx4_reverse(ijkl)
+
+    def test_idx4_reverse_all(self, n=10000, nmax=(1 << 63) - 1):
+        for ijkl in random.sample(range(nmax), k=n):
+            self.check_idx4_reverse_all(ijkl)
+
+    def test_canonical_idx4(self, n=10000, nmax=(1 << 63) - 1):
+        for ijkl in random.sample(range(nmax), k=n):
+            self.check_canonical_idx4(ijkl)
+
+    def test_compound_idx4_reverse_is_canonical(self, n=10000, nmax=(1 << 63) - 1):
+        for ijkl in random.sample(range(nmax), k=n):
+            self.check_compound_idx4_reverse_is_canonical(ijkl)
```

Comment:
@TApplencourt would you prefer these to be combined?
(we could also include the 2idx one, but the name of the indexing variable will be less descriptive)
```suggestion
    def test_idx4(self, n=10000, nmax=(1 << 63) - 1):
        for ijkl in random.sample(range(nmax), k=n):
            self.check_idx4_reverse(ijkl)
            self.check_idx4_reverse_all(ijkl)
            self.check_canonical_idx4(ijkl)
            self.check_compound_idx4_reverse_is_canonical(ijkl)
```

Reply:
Unit test who are unit are nice :)